File: frontend/src/services/history.ts
import api from './api';
import type { HistoryResponse, HistoryStats } from '../types/history';

interface HistoryFilters {
  topic?: string;
  channel?: string;
  from?: string;
  to?: string;
  sort?: string;
  order?: 'asc' | 'desc';
  page?: number;
  limit?: number;
}

export async function getHistory(filters: HistoryFilters = {}): Promise<HistoryResponse> {
  const params = new URLSearchParams();
  if (filters.topic) params.set('topic', filters.topic);
  if (filters.channel) params.set('channel', filters.channel);
  if (filters.from) params.set('from', filters.from);
  if (filters.to) params.set('to', filters.to);
  if (filters.sort) params.set('sort', filters.sort);
  if (filters.order) params.set('order', filters.order);
  if (filters.page) params.set('page', String(filters.page));
  if (filters.limit) params.set('limit', String(filters.limit));

  const { data } = await api.get<HistoryResponse>(`/history?${params.toString()}`);
  return data;
}

export async function getHistoryStats(): Promise<HistoryStats> {
  const { data } = await api.get<HistoryStats>('/history/stats');
  return data;
}

File: api/routers/history.py
"""Research history endpoints."""

from __future__ import annotations

from datetime import datetime

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from api.dependencies import get_db, verify_api_key
from api.models.schemas.history import HistoryListResponse, HistoryStatsResponse
from api.services import history_service

router = APIRouter(prefix="/api/history", tags=["history"], dependencies=[Depends(verify_api_key)])


@router.get("", response_model=HistoryListResponse)
async def list_history(
    topic: str | None = Query(None),
    channel: str | None = Query(None),
    date_from: datetime | None = Query(None, alias="from"),
    date_to: datetime | None = Query(None, alias="to"),
    sort: str = Query("researched_at"),
    order: str = Query("desc"),
    page: int = Query(1, ge=1),
    limit: int = Query(50, ge=1, le=200),
    db: AsyncSession = Depends(get_db),
):
    total, items = await history_service.list_history(
        db,
        topic=topic,
        channel=channel,
        date_from=date_from,
        date_to=date_to,
        sort=sort,
        order=order,
        page=page,
        limit=limit,
    )
    return {"total": total, "page": page, "limit": limit, "items": items}


@router.get("/stats", response_model=HistoryStatsResponse)
async def history_stats(db: AsyncSession = Depends(get_db)):
    return await history_service.get_history_stats(db)
